polyglot/piranha/src/piranha.rs

```diff
@@ -59,76 +72,146 @@ impl SourceCodeUnit {
-        if let Some(edit) = Rule::get_edit_for_context(
-          &self.clone(),
-          current_edit.start_byte,
-          current_edit.new_end_byte,
-          rules_store,
-          &next_rules_by_scope[PARENT],
-        ) {
-          #[rustfmt::skip]
-          info!( "{}", format!( "Cleaning up the context, by applying the rule - {}", edit.matched_rule().name()).green());
-          // Apply the matched rule to the parent
-          current_edit = self.apply_edit(&edit, parser);
-          current_rule = edit.matched_rule();
-          // Add the (tag, code_snippet) mapping to substitution table.
-          self.add_to_substitutions(edit.matches());
-        } else {
-          // No more parents found for cleanup
-          break;
-        }
+    // When rule is a "rewrite" rule :
+    // Update the first match of the rewrite rule
+    // Add mappings to the substitution
+    // Propagate each applied edit. The next rule will be applied relative to the application of this edit. 
+    if !rule.is_match_only_rule() {
+      if let Some(edit) = rule.get_edit(&self.clone(), rule_store, scope_node, true) {
+        self.rewrites_mut().push(edit.clone());
+        query_again = true;
+
+        // Add all the (code_snippet, tag) mapping to the substitution table.
+        self.add_to_substitutions(edit.matches());
+
+        // Apply edit_1
+        let applied_ts_edit = self.apply_edit(&edit, parser);
+
+        self.propagate(applied_ts_edit, rule, rule_store, parser);
+      }
+    }
+    // When rule is a "match-only" rule :
+    // Get all the matches
+    // Add mappings to the substitution
+    // Propagate each match. Note that,  we pass a identity edit (where old range == new range) in to the propagate logic. 
+    // The next edit will be applied relative to the identity edit. 
+    else {
+      for m in rule.get_matches(&self.clone(), rule_store, scope_node, true) {
+        self.matches_mut().push((rule.clone(), m.clone()));
+
+        self.add_to_substitutions(m.matches());
+        
+        // Create identity edit, that replaces a range with itself. 
+        let identity_edit = InputEdit {
+          start_byte: m.range().start_byte,
+          old_end_byte: m.range().end_byte,
+          new_end_byte: m.range().end_byte,
+          start_position: m.range().start_point,
+          old_end_position: m.range().end_point,
+          new_end_position: m.range().end_point,
+        };
+
+        self.propagate(identity_edit, rule.clone(), rule_store, parser);
+      }
+    }
+    query_again
+  }
+
+  /// This is the propagation logic of the Piranha's main algorithm.
+  /// Parameters:
+  ///  * Rule `rule` that was just applied
+  ///  * `applied_ts_edit` -  it's(`rule`'s) application site (in terms of replacement range)
+  ///  * `rule_store` - contains the input "rule graph"
+  ///
+  /// Algo:
```

Comment:
Is this by convention? I don't see anything in [RFC1574](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.htm). Probably just use "Algorithm"

Reply:
Fixed.